### PR TITLE
Google

### PR DESCRIPTION
**Pull request by lostinthestory**
*Fixed priority of reading parameters - Node config now takes precedence over msg.XXXX*

### Google Geocoding

* Added node for Google Geocoding that allows geocoding and reverse geocoding requests to be handled

### Google Places

* Added support in Google Places node for Places Nearby and Places Details requests
* Modified node to allow for API Key (key) to be passed in, rather than be required in the node config.
* Added support for additional optional parameters to be input
* Parameters can either be passed in (via msg.XXXX) or configured in the node itself. **Node will take parameters configured in the node at higher priority than those passed in via msg.XXXX.** ~~Node will take parameters passed via msg.XXXX at higher priority than those configured in the node.~~
* Updated html file to reflect all available optional params.

### CODE REVIEW

**Comment by hindessm** [comment]
I assume the API key for geocoding is the same type of key as the one used by google places? If so, then it should use a ```google-api-config```  node (see places.{js,html} for example of usage) rather than the key directly so that multiple nodes in a flow can share credentials.

Aside: you can add commits [or modify them with rebase -i] on your branch and this pull request will reflect the changes you push so you don't need to create a new pull request to make changes.

File: google/geocoding.html [hunk 1]
```diff
@@ -0,0 +1,133 @@
+<!--
+  Copyright 2014 IBM Corp.
```
Comment (hindessm) [inline google/geocoding.html]: The copyright year for new files should be 2015 not 2014.

**Comment by hindessm** [comment]
Sorry I wasn't clear enough but the copyright year(s) for new files should be "Copyright 2015 IBM Corp." but the copyright for updated files should contain the oldest year and most recent year - such as "Copyright 2014, 2015 IBM Corp.". The ```geocoding.html``` file still contains 2014 not 2015.

I notice the code is now corrected with respect to "lng" being changed to "lon" but ```.html``` files still contain references to "lng" in the documentation sections.

For consistency with other nodes, the credentials should probably be first in the ui dialogs for node configuration. There might be other consistency issues in the ui but I've not looked at this in more detail yet.

The logic for the lat/lon parameters is incorrect. The old code use ```msg.location.hasOwnProperty("lat")``` rather than simply a logic test for ```msg.location.lat``` since 0 is false but also a valid latitude. It so happens that one of the tests has a longitude of 0 and is currently failing. You use the logic value of lat/lon in a number of places; these need to be fixed.

I suspect there are other logic value errors. For instance, I suspect 0 is a valid minprice and/or maxprice. Care needs to be taken when checking both the node and msg parameters.

The old code used a default for the radius if it was not set. The new code raises an error. We need to maintain backwards compatibility (and I prefer the more helpful old behaviour anyway).

The old code sets the value of radius (if it was configured) to max(radius, 50000) rather than throwing an error if it was greater than the 50000 maximum. We need to maintain backwards compatibility. The results still meet the requirement of being within the user supplied radius. It is slightly confusing in that the no results case may not really mean there are no results within the supplied radius but overall I think it is helpful.

The new code removes the ```msg.location.description``` field that was previously constructed from ```name + ", " + address```. This should be put back to maintain backwards compatibility and meet the msg conventions. It is fine to add the ```msg.location.address``` property (and perhaps this convention should be considered for addition to the documentation). However, you will need to update the two tests to either be more flexible about only checking for the required ```msg.location``` properties or to expect the new ```msg.location.address``` property.

I think if you fix the above backward compatibility issues and make the test updates for ```msg.location.address``` then the second and third tests should pass again.

I notice the code for zero results case still does a node.warn and sets node.status to failed. You need to move this up to the ```if (msg.statusCode !== 200) {``` block.

The node.warn for the request error cases should probably also contain the human readable error test from body.error_message (if set) as it did in the old code. The fourth test should pass again if you fix this.

The convention for code indentation is 4 spaces. This code is using tabs which are equivalent to 8 spaces.

The project convention is also to squash commits so that the project history doesn't contain too many trivial fixes but instead one or more commits to reflect the functional changes.

**Comment by lostinthestory** [comment]
* Will make the trivial fixes today. Some were already done, and just not pushed, as I caught them testing afterwards.

* Will edit logic to ensure 0 is not being ignored.

* As for the node.warn in the zero results case, I had actually moved that because it seemed your code was also warning for zero results
<code>} else if (data.results.length < 1) {
                    node.warn("no results returned");
                    node.status({fill:"red",shape:"ring",text:"failed"});
                    delete msg.payload;
                    msg.error = "no results returned";
                    msg.data = data;
                }</code>
Will change this to not warn, as it is a valid case.

**Comment by lostinthestory** [comment]
Oddly enough, I had local tests passing when setting msg.location.lon to 0. It seems to still ready the value if I set it via the HTTP request or directly in the node, even though both of the statements should evaluate false. Changing the logic to be correct all the time. Just thought that was interesting.

**Comment by hindessm** [comment]
Indeed. The string "0" is true but if the location was set via, for example, a function node it might be integer and thus false. As you say it is easy to write a test that happens to work. (The delay node had a related issue recently with a number being a string when configured through the UI but an integer in the test case. The result was a passing test but a '+' being interpreted as concatenation rather than addition in real usage.)

**Comment by lostinthestory** [comment]
Tweaked the tests a bit to be more forgiving if additional data is returned. They are passing now. 
 google places
    query node
  ✓ can be loaded without credentials 
  ✓ should return lat/lon from postcode 
  ✓ should return place in vicinity 
  ✓ warns about invalid API key 
Remaining failures are unrelated files.
Will add more tests to the file to test the new functionality.

**Comment by hindessm** [comment]
You need to rebase on to the current master then you will have the changes that fix the tests. You should probably squash some of the commits too for clarity of project history.

**Comment by lostinthestory** [comment]
Squashed commits into a single commit, and ensured all missing commits from master are pulled into this branch. When comparing to node-red's master, my files are the only diff, however the other tests seem to still be failing, even when I run the tests local on just the master branch.

**Pull request closed without merging**